compiler/src/main/kotlin/edu/cornell/cs/apl/viaduct/codegeneration/AbstractCodeGenerator.kt

```diff
@@ -1,17 +1,39 @@
 package edu.cornell.cs.apl.viaduct.codegeneration
 
+import com.squareup.kotlinpoet.BOOLEAN
 import com.squareup.kotlinpoet.CodeBlock
+import com.squareup.kotlinpoet.INT
+import com.squareup.kotlinpoet.STRING
+import com.squareup.kotlinpoet.TypeName
+import com.squareup.kotlinpoet.U_BYTE_ARRAY
+import edu.cornell.cs.apl.viaduct.errors.CodeGenerationError
+import edu.cornell.cs.apl.viaduct.errors.RuntimeError
+import edu.cornell.cs.apl.viaduct.selection.CommunicationEvent
+import edu.cornell.cs.apl.viaduct.syntax.Arguments
+import edu.cornell.cs.apl.viaduct.syntax.ClassNameNode
 import edu.cornell.cs.apl.viaduct.syntax.Protocol
+import edu.cornell.cs.apl.viaduct.syntax.datatypes.ImmutableCell
+import edu.cornell.cs.apl.viaduct.syntax.datatypes.MutableCell
+import edu.cornell.cs.apl.viaduct.syntax.datatypes.Vector
 import edu.cornell.cs.apl.viaduct.syntax.intermediate.AtomicExpressionNode
 import edu.cornell.cs.apl.viaduct.syntax.intermediate.DeclarationNode
+import edu.cornell.cs.apl.viaduct.syntax.intermediate.ExpressionNode
 import edu.cornell.cs.apl.viaduct.syntax.intermediate.LetNode
+import edu.cornell.cs.apl.viaduct.syntax.intermediate.OutParameterConstructorInitializerNode
+import edu.cornell.cs.apl.viaduct.syntax.intermediate.OutParameterExpressionInitializerNode
 import edu.cornell.cs.apl.viaduct.syntax.intermediate.OutParameterInitializationNode
 import edu.cornell.cs.apl.viaduct.syntax.intermediate.OutputNode
 import edu.cornell.cs.apl.viaduct.syntax.intermediate.SimpleStatementNode
 import edu.cornell.cs.apl.viaduct.syntax.intermediate.UpdateNode
-
-abstract class AbstractCodeGenerator : CodeGenerator {
-
+import edu.cornell.cs.apl.viaduct.syntax.types.BooleanType
+import edu.cornell.cs.apl.viaduct.syntax.types.ByteVecType
+import edu.cornell.cs.apl.viaduct.syntax.types.IntegerType
+import edu.cornell.cs.apl.viaduct.syntax.types.StringType
+import edu.cornell.cs.apl.viaduct.syntax.types.ValueType
+import edu.cornell.cs.apl.viaduct.syntax.values.Value
+
+abstract class AbstractCodeGenerator(val context: CodeGeneratorContext) : CodeGenerator {
+    val runtimeErrorClass = RuntimeError::class
```

Comment:
`RuntimeError` is part of the `compile` module. Generated code cannot refer to this. You should use `EquivocationException` for this. Even better, use the `assertEquals` function from the same file.

Reply:
Who is the expectedValueProvider that I should pass into the EquivocationException here? In the case where we go from Replication ->  Local, it isnt clear to me who the expected value provider is

Reply:
It's whomever sent you the first value in the event set. Example: if you receive a value from Alice, Bob, and Chuck, you compare Bob and Chuck's values to Alice's value (but don't directly compare Bob's value to Chuck's value). In this case, expected value is Alice's value, expected provider is Alice, actual value is Bob's value, and actual provider is Bob. Same for Chuck.